WIP: Add decorators for API endpoints

HarrySky opened the pull request:
This PR will introduce decorators to all API endpoints. Possibly also introduce new methods into `process` module. **Work In Progress**, I will let you know when it is done. Tests for async flow will come in another PR if you don't mind. I suggest squashing this PR into one commit when it is done (to reduce noise in commit history)

HarrySky commented on tekore/client/process.py:
```diff
@@ -24,3 +24,12 @@ def model_list(type_: type, top_item: str):
         return ModelList(type_(**i) for i in json[top_item])
     return post_func
+
+def cursor_paging(type_: type, top_item: str):
```
This name can be changed to something more generic. Right now it is used for parsing into `FullArtistCursorPaging` object in `follow.followed_artists`. Up for a discussion

felix-hilden replied:
Perhaps I would change it. It isn't really tied to the idea of paging as much as accessing an item and using that to unpack, as you already mentioned in the docstring.

HarrySky replied:
In new commit I modified `single` to accept `top_item`. I think this way is better, what do you think?

felix-hilden replied:
I agree!

HarrySky commented on tekore/client/process.py:
```diff
@@ -25,5 +25,5 @@ def model_list(type_: type, top_item: str):
     """
     def post_func(json: dict):
-        return ModelList(type_(**i) for i in json[top_item])
+        return ModelList(type_(**i) if i is not None else None for i in json[top_item])
```
Added check for `None` in `model_list` which is performed in `track.tracks_audio_features`. Right now I don't see problems, but maybe there should be separate method in `process` which explicitly checks for None, since modified method silently converts to `None`, while original method would raise if value was `None`.

felix-hilden replied:
It's probably fine like that. Filtering Nones out would be wrong, but this preserves them and does not lead to confusion.

HarrySky commented on tekore/client/process.py:
```diff
@@ -27,3 +33,18 @@ def model_list(type_: type, top_item: str):
+    """
+    Unpack search result dicts into respective paging type constructors.
+    """
+    return tuple(paging_type[key[:-1]](**json[key]) for key in json.keys())
```
One thought here: maybe this code should go to `api/search.py`, since it is too specific and used only for processing search result.

felix-hilden replied:
Most definitely. I did a similar thing for paging 👍

HarrySky commented:
For some reason tests do not cover lines below in `browse.recommendations` method (`tekore/client/api/browse/__init__.py`):

```python
if artist_ids is not None:
    params['seed_artists'] = ','.join(artist_ids)
```

Bringing it to your attention, continuing working.

felix-hilden commented:
Yup, the statement is covered, but I think it is because it is always evaluated as true. Line coverage is 100% but `artist_ids` is never None.

HarrySky commented:
Well, done! :tada: Can you please review the changes, I added and changed code in `process` module. Tests are running without errors right now. Coverage is at 97%, but I was not planning to write tests in this PR if this is okay

felix-hilden commented on tekore/client/base.py:
```diff
@@ -121,25 +121,54 @@ def _request(
+            headers=self._create_headers(content_type),
             params=parse_url_params(params),
-            data=json.dumps(payload) if payload is not None else None
+            data=data
```
Is the JSON dump not required? That's neat if it's true.

HarrySky replied:
It is dumped above if `content-type` is JSON. I had to introduce it because of errors that I encountered when working on playlist image adding endpoint

felix-hilden replied:
Oh, I missed it! Please see my review on the whole change.

HarrySky commented:
Yeah, you are right. I will rewrite it the other way soon, for today I'm done :)

If you have any other ideas/proposals for this PR - let me know, I will try to finish it by Friday.

felix-hilden commented:
It all looks good to me, thank you! If there's something small I'll change it later. It's good to know that the current tests pass well!

HarrySky commented:
I have removed `content_type` and just create and return `Request` now from that playlist image changing endpoint. All commits are squashed into one and ready to be merged. Let me know if there is something else needed to be done :)

HarrySky commented:
Travis is failing currently because there is a mistake in `setup.py`:

`httpx=0.11.*` should be `httpx==0.11.*`

But I think it should not be fixed in this PR.

felix-hilden commented:
Right, thank you! I'll fix it in the main branch.
